graphql_compiler/schema/__init__.py

```diff
@@ -340,36 +340,47 @@ def _serialize_date(value: Any) -> str:
 
 def _parse_date_value(value: Any) -> date:
     """Deserialize a Date object from its proper ISO-8601 representation."""
-    return arrow.get(value, "YYYY-MM-DD").date()
+    if type(value) == date:
+        # We prefer exact type equality instead of isinstance() because datetime objects
+        # are subclasses of date but are not interchangeable for dates for our purposes.
+        return value
+    elif isinstance(value, str):
+        return arrow.get(value, "YYYY-MM-DD").date()
+    else:
+        raise ValueError(
+            f"Expected a date object or its ISO-8601 string representation. "
+            f"Got {value} of type {type(value)} instead."
+        )
 
 
 def _serialize_datetime(value: Any) -> str:
```

Comment:
Unfortunately we have to follow the upstream contract, since this function is going to get assigned to a GraphQL type object's `serialize()` slot. The upstream contract is that both the `parse()` and the `serialize()` functions accept `Any`:
https://github.com/graphql-python/graphql-core/blob/master/src/graphql/type/scalars.py#L105

Hence the runtime type checks.